src/error.rs

```diff
@@ -5,78 +5,119 @@
+/// either the underlying OS or a custom error reason.
 ///
-/// This type is small and no-std compatible.
+/// The [`Error::raw_os_error()`] will indicate if the error is from the OS, and
+/// if so, which error code the OS gave the application. If such an error is
+/// encountered, please consult with your system documentation.
 #[derive(Copy, Clone, Eq, PartialEq)]
-pub struct Error(pub(crate) NonZeroU32);
+pub struct Error(NonZeroU32);
 
+// This NonZeroU32 has enough room to store 3 types of values:
+//   - OS Errors:         in range [1, 1 << 31) (i.e. positive i32 values)
+//   - Custom u16 Errors: in range [1 << 31, 1 << 31 + 1 << 16)
+//   - Unknown Errors:    currently just (1 << 32) - 1
+// TODO simplify impls with try_from when version >= 1.34
 impl Error {
     /// An unknown error.
-    pub const UNKNOWN: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNKNOWN) });
+    pub(crate) const UNKNOWN: Error = Self(unsafe { NonZeroU32::new_unchecked(u32::max_value()) });
+    const CUSTOM_START: u32 = 1 << 31;
 
-    /// No generator is available.
-    pub const UNAVAILABLE: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNAVAILABLE) });
+    pub(crate) fn from_os_error(errno: i32) -> Self {
+        if errno > 0 {
+            Self(NonZeroU32::new(errno as u32).unwrap())
+        } else {
+            Self::UNKNOWN
+        }
+    }
 
-    /// Extract the error code.
-    ///
-    /// This may equal one of the codes defined in this library or may be a
-    /// system error code.
-    ///
-    /// One may attempt to format this error via the `Display` implementation.
-    pub fn code(&self) -> NonZeroU32 {
-        self.0
+    pub(crate) fn from_custom_error(custom: u16) -> Self {
+        Self(NonZeroU32::new(custom as u32 + Self::CUSTOM_START).unwrap())
     }
 
-    pub(crate) fn msg(&self) -> Option<&'static str> {
-        if let Some(msg) = crate::imp::error_msg_inner(self.0) {
-            Some(msg)
+    /// Extract the raw OS error code (if this error came from the OS)
+    ///
+    /// This method is identical to `std::io::Error::raw_os_error()`, except
+    /// that it works in `no_std` contexts. If this method returns `None`, the
+    /// error value can still be formatted via the `Diplay` implementation.
+    pub fn raw_os_error(&self) -> Option<i32> {
+        if self.0.get() < Self::CUSTOM_START {
+            Some(self.0.get() as i32)
         } else {
-            match *self {
-                Error::UNKNOWN => Some("getrandom: unknown error"),
-                Error::UNAVAILABLE => Some("getrandom: unavailable"),
-                _ => None,
-            }
+            None
         }
     }
-}
 
-impl fmt::Debug for Error {
-    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        match self.msg() {
-            Some(msg) => write!(f, "Error(\"{}\")", msg),
-            None => write!(f, "Error(0x{:08X})", self.0),
+    pub fn custom_code(&self) -> Option<u16> {
```

Comment:
Undocumented public fn.

One thing seems incongruous to me: we have `2<<31 - 1` error codes available, but only allow use of 2<<16 of these (labelled 'custom'). I think the issue is just terminology: these codes are for "getrandom internal" usage.

Perhaps we should use values from `2<<32 - 2<<16` as "special/custom codes"?

Reply:
I reworked how a user gets the error information from an `Error` by introducing an `ErrorKind` enum (and corresponding `Error::kind()` method). This enum represents the three sources of errors:

- `Os` errors (positive `i32` values)
- `Internal` errors (`u16` values)
- `Custom` errors consisting on two parts:
  - `block` - A positive `i16` used to group errors together
  - `code` - The actual `u16` error code within a block.
- In a sense, our `Internal` errors are `Custom` errors in "block 0".